### DISCUSSION TITLE
Unexpected behaviour working with `JITParticle`s and custom kernels

### DISCUSSION THREAD

**Discussion by VeckoTheGecko** [Parcels-code/Parcels#1232] [Q&A] [answered]
I've been encountering very strange behaviour when using the `JITParticle` combined with custom kernels.

Eg.
- Using a simple nested `if` statement in a kernel causes a bugged output file (part of the if statement getting hit regardless of the condition, causing all particles to change instantly), while collapsing the nested `if` statement to an equivalent `<condition> and <condition>` causes no issues
- Memory access violation bug appearing out of nowhere, despite not changing the kernels (duplicate of #793)

I guess my question is more to get a feel of:
- whether these bugs are known/encountered by the developers often
- whether these bugs/unexpected behaviour is something that can be attributed to parcels, or is it moreso the dependencies/way that things are compiled to C
- whether working on Windows makes me more prone to running into these bugs (and if so, how much worse is it to work on Windows compared to Linux? Is it worth swinging across?)

PS:
I experimented using the `ScipyParticle`, and my simulation went from 2 minutes (100% completion) to 2hr 30mins (22% completion) so I don't think thats an option for me. Its "`JITParticle` or nothing" with my current kernels.

**Comment by erikvansebille** [comment, marked as answer]
Hi @VeckoTheGecko, comments to your two questions below

> Using a simple nested if statement in a kernel causes a bugged output file (part of the if statement getting hit regardless of the condition, causing all particles to change instantly), while collapsing the nested if statement to an equivalent <condition> and <condition> causes no issues

Hmm, nested if-statements are not explicitly implemented in the codeconverter (which converts the python-code in Kernels to JIT-code. So I wouldn't be surprised that this goes wrong. I also don't know how to easily fix this, not how to give an explicit warning. I'll open an Issue and think about it. But much safer to use non-nested-if-statements!

> Memory access violation bug appearing out of nowhere, despite not changing the kernels (duplicate of https://github.com/OceanParcels/parcels/issues/793)

I have never experienced this Issue myself, but don't run Windows. I also haven't seen it from anyone else here in my group (none of whom use Windows). So he bug seems to be explicitly related to Windows? No idea how to even begin to fix this...
